File: src/lib/providers/anime-episode-build.ts
import { pickEpisodeTitle, type AniZipMapping } from "@/lib/providers/anizip";
import type { AnimeKitsuMeta } from "@/lib/providers/anime-kitsu-addon";
import type { KitsuEpisode } from "@/lib/providers/kitsu";
import type { TvdbEpisode } from "@/lib/providers/tvdb";

export function buildKitsuEpisodes(
  addonMeta: AnimeKitsuMeta | null,
  kitsuRawEpisodes: KitsuEpisode[],
): KitsuEpisode[] {
  if (!addonMeta?.videos || addonMeta.videos.length === 0) return kitsuRawEpisodes;
  const kitsuById = new Map<number, KitsuEpisode>();
  for (const ep of kitsuRawEpisodes) kitsuById.set(ep.number, ep);
  return addonMeta.videos.map((v): KitsuEpisode => {
    const k = kitsuById.get(v.episode);
    return {
      id: k?.id ?? v.episode,
      number: v.episode,
      seasonNumber: v.season ?? 1,
      title: v.title || k?.title || `Episode ${v.episode}`,
      synopsis: v.overview ?? k?.synopsis ?? "",
      thumbnail: v.thumbnail ?? k?.thumbnail ?? null,
      airdate: v.released ?? k?.airdate ?? null,
      length: k?.length ?? null,
      streamId: v.id,
      imdbId: v.imdb_id,
      imdbSeason: v.imdbSeason,
      imdbEpisode: v.imdbEpisode,
    };
  });
}

export function mergeAniZipEpisodes(episodes: KitsuEpisode[], aniZip: AniZipMapping | null): void {
  if (!aniZip?.episodes) return;
  const azImdb = aniZip.mappings?.imdb_id;
  for (const ep of episodes) {
    const az = aniZip.episodes[String(ep.number)];
    if (!az) continue;
    const enrichedTitle = pickEpisodeTitle(az);
    if (enrichedTitle && (!ep.title || ep.title === `Episode ${ep.number}`)) {
      ep.title = enrichedTitle;
    }
    if (az.overview && !ep.synopsis) ep.synopsis = az.overview;
    if (az.image) {
      if (ep.thumbnail && ep.thumbnail !== az.image && !ep.thumbnailFallback) {
        ep.thumbnailFallback = ep.thumbnail;
      }
      ep.thumbnail = az.image;
    }
    if (az.airDate) ep.airdate = az.airDate;
    if (az.runtime && !ep.length) ep.length = az.runtime;
    if (az.filler) ep.filler = true;
    if (az.absoluteEpisodeNumber) ep.absoluteNumber = az.absoluteEpisodeNumber;
    if (az.tvdbId) ep.tvdbEpisodeId = az.tvdbId;
    if (ep.rating == null && az.rating != null) {
      const r = Number(az.rating);
      if (Number.isFinite(r) && r > 0) {
        if (!ep.airdate || new Date(ep.airdate).getTime() <= Date.now()) {
          ep.rating = r;
        }
      }
    }
    if (az.seasonNumber != null && az.seasonNumber > 0 && az.episodeNumber != null) {
      if (azImdb) ep.imdbId = azImdb;
      if (ep.imdbSeason == null) ep.imdbSeason = az.seasonNumber;
      if (ep.imdbEpisode == null) ep.imdbEpisode = az.episodeNumber;
    }
  }
}

export function mergeTvdbEpisodes(episodes: KitsuEpisode[], tvdbEps: TvdbEpisode[] | null): void {
  if (!tvdbEps || tvdbEps.length === 0) return;
  const tvdbById = new Map<number, TvdbEpisode>();
  const tvdbByAbsolute = new Map<number, TvdbEpisode>();
  const tvdbBySeasonAndEpisode = new Map<string, TvdbEpisode>();

  for (const e of tvdbEps) {
    tvdbById.set(e.id, e);
    if (e.absoluteNumber != null) tvdbByAbsolute.set(e.absoluteNumber, e);
    tvdbBySeasonAndEpisode.set(`${e.seasonNumber}:${e.number}`, e);
  }

  for (const ep of episodes) {
    let tvdbEp: TvdbEpisode | undefined;

    if (ep.tvdbEpisodeId) tvdbEp = tvdbById.get(ep.tvdbEpisodeId);
    if (!tvdbEp && ep.absoluteNumber) tvdbEp = tvdbByAbsolute.get(ep.absoluteNumber);
    if (!tvdbEp && ep.imdbSeason != null && ep.imdbEpisode != null) {
      tvdbEp = tvdbBySeasonAndEpisode.get(`${ep.imdbSeason}:${ep.imdbEpisode}`);
    }
    if (!tvdbEp) tvdbEp = tvdbBySeasonAndEpisode.get(`${ep.seasonNumber}:${ep.number}`);
    if (!tvdbEp) tvdbEp = tvdbByAbsolute.get(ep.number);

    if (tvdbEp) {
      if (tvdbEp.name && (!ep.title || ep.title === `Episode ${ep.number}`)) {
        ep.title = tvdbEp.name;
      }
      if (tvdbEp.aired) ep.airdate = tvdbEp.aired;
      if (tvdbEp.overview && !ep.synopsis) ep.synopsis = tvdbEp.overview;
      if (tvdbEp.image) {
        if (ep.thumbnail && ep.thumbnail !== tvdbEp.image && !ep.thumbnailFallback) {
          ep.thumbnailFallback = ep.thumbnail;
        }
        ep.thumbnail = tvdbEp.image;
      }
      if (tvdbEp.runtime && !ep.length) ep.length = tvdbEp.runtime;
    }
  }
}

File: clients/lib/domain/anime/anime_episode_build.dart
import 'anime_kitsu_addon.dart';
import 'anizip.dart';
import 'kitsu_client.dart';

/// Builds the episode list from the addon meta's videos when present — carrying
/// their stream and IMDb cross-ids — and otherwise returns the raw Kitsu
/// episodes unchanged. The Kitsu episode of the same number supplies fallback
/// title, synopsis, thumbnail, airdate and length. Ported 1:1 from
/// `buildKitsuEpisodes`.
List<KitsuEpisode> buildKitsuEpisodes(
  AnimeKitsuMeta? addonMeta,
  List<KitsuEpisode> kitsuRawEpisodes,
) {
  final videos = addonMeta?.videos;
  if (videos == null || videos.isEmpty) return kitsuRawEpisodes;
  final kitsuByNumber = <int, KitsuEpisode>{
    for (final ep in kitsuRawEpisodes) ep.number: ep,
  };
  return [for (final v in videos) _fromVideo(v, kitsuByNumber[v.episode])];
}

KitsuEpisode _fromVideo(AnimeKitsuVideo v, KitsuEpisode? k) {
  final title = v.title.isNotEmpty
      ? v.title
      : (k != null && k.title.isNotEmpty ? k.title : 'Episode ${v.episode}');
  return KitsuEpisode(
    id: k?.id ?? v.episode,
    number: v.episode,
    seasonNumber: v.season ?? 1,
    title: title,
    synopsis: v.overview ?? k?.synopsis ?? '',
    thumbnail: v.thumbnail ?? k?.thumbnail,
    airdate: v.released ?? k?.airdate,
    length: k?.length,
    streamId: v.id,
    imdbId: v.imdbId,
    imdbSeason: v.imdbSeason,
    imdbEpisode: v.imdbEpisode,
  );
}

/// Enriches [episodes] in place from an ani.zip mapping: better titles, missing
/// overviews/thumbnails/airdates/runtimes, filler and absolute-number flags,
/// ratings, and IMDb season/episode routing derived from the mapping. Only
/// fills gaps — an existing value is never overwritten, except the airdate,
/// which ani.zip always wins. Ported 1:1 from `mergeAniZipEpisodes`.
void mergeAniZipEpisodes(List<KitsuEpisode> episodes, AniZipMapping? aniZip) {
  if (aniZip == null) return;
  final byNumber = aniZip.episodes;
  final azImdb = aniZip.mappings?.imdbId;
  for (final ep in episodes) {
    final az = byNumber['${ep.number}'];
    if (az == null) continue;

    final enrichedTitle = pickEpisodeTitle(az);
    if (enrichedTitle != null &&
        enrichedTitle.isNotEmpty &&
        (ep.title.isEmpty || ep.title == 'Episode ${ep.number}')) {
      ep.title = enrichedTitle;
    }
    if (az.overview.isNotEmpty && ep.synopsis.isEmpty) {
      ep.synopsis = az.overview;
    }
    final image = az.image;
    if (image != null &&
        image.isNotEmpty &&
        (ep.thumbnail == null || ep.thumbnail!.isEmpty)) {
      ep.thumbnail = image;
    }
    final airDate = az.airDate;
    if (airDate != null && airDate.isNotEmpty) ep.airdate = airDate;
    final runtime = az.runtime;
    if (runtime != null &&
        runtime != 0 &&
        (ep.length == null || ep.length == 0)) {
      ep.length = runtime;
    }
    if (az.filler == true) ep.filler = true;
    final abs = az.absoluteEpisodeNumber;
    if (abs != null && abs != 0) ep.absoluteNumber = abs;
    if (ep.rating == null && az.rating != null) {
      final r = num.tryParse(az.rating!);
      if (r != null && r.isFinite && r > 0) ep.rating = r;
    }
    final azSeason = az.seasonNumber;
    if (azSeason != null && azSeason > 0 && az.episodeNumber != null) {
      if (azImdb != null) ep.imdbId = azImdb;
      ep.imdbSeason ??= azSeason;
      ep.imdbEpisode ??= az.episodeNumber;
    }
  }
}
